### ISSUE TITLE
[OAuth2] return 401 instead of 302 on xhr requests

### ISSUE THREAD

**Issue by AurelienBegou** [envoyproxy/envoy#30102]
*Title*: [OAuth2] return 401 instead of 302 on xhr requests

*Description*:
In the context of Single Page Application (SPA) the code is stored in cache in the browser and at the end of the session having a 302 redirect will cause issue in the design. Typically when xhr request is detected it should return 401 instead.

Is it such configuration possible ? I try to search on the documentation but was not able to found it.

[optional *Relevant Links*:]
Example on another auth-proxy
Oauth2-Proxy: https://oauth2-proxy.github.io/oauth2-proxy/docs/behaviour/
Specific line:
"If authentication is required but missing then the user is asked to log in and redirected to the authentication provider (unless it is an Ajax request, i.e. one with Accept: application/json, in which case 401 Unauthorized is returned)"

Thanks for the help

**Comment by kyessenov** [comment]
Seems reasonable. CC @derekargueta

**Comment by AurelienBegou** [comment]
> Seems reasonable. CC @derekargueta

Thanks, here an additional implementation to force behavior regarding xhr/AJAX requests https://cloud.google.com/iap/docs/sessions-howto#understanding_the_response

**Comment by samo1** [comment]
@kyessenov I would like to re-open this issue if possible.

This requested feature would be quite useful in our use case. I am testing a POC solution where I have added a new configuration option to oauth2 filter `ajax_request_matcher` which does not allow redirect of matching requests and returns 401 Unauthorized immediately. It seems to be working well so far. The changes I made - https://github.com/samo1/envoy/compare/main...oauth2_ajax_matcher

I am willing to open a merge request if we agree on a solution. Thanks.
